### DISCUSSION TITLE
Export to PDF report displays incorrect HTML content for Arabic tables

### DISCUSSION THREAD

**Discussion by prshnt0071** [eclipse-birt/birt#1114] [Q&A] [answered]
We have input made by user with huge tables and bullets through rich text editor stored in the database as an HTML content where the labels are shown to the right and text area to the left.

The BIRT report displays the HTML content in the report through the web viewer correctly where the table columns are displayed to right and text against those to the left, but, when we export this to PDF using the Report viewer, the document shows the table columns to the left which was seen perfectly on the BIRT viewer.

It would be really great if I can get any help on the same urgently or else can forward me to a relevant thread that already has been resolved.
Below is the actual exported PDF output of that HTML content where the table headers are displayed to the left as marked in blue.

![image](https://user-images.githubusercontent.com/118506393/202658282-36eebcc5-3f65-4e7e-91c9-3eba48108f10.png)

Expected Output:
![image](https://user-images.githubusercontent.com/118506393/202659721-fce911e7-406f-4b37-8e58-8793580d50a3.png)

**Comment by wimjongman** [comment]
You say "urgently." Do you want to hire one of us as a consultant to fix this for you?

**Reply by prshnt0071** [reply]
We have tried all the possible solution and because of this issue, the complete project is on risk...Grateful, if a help can be lended on the same.

**Reply by wimjongman** [reply]
You did not answer the question.

**Reply by prshnt0071** [reply]
No, I am not here for hiring. Being such a big community, I assume that someone must have encountered similar problem and I could find answers to this.

**Reply by wimjongman** [reply]
That is fine. I thought I ask because you said it is urgent and the complete project is at risk.

For anyone to help you. We need more information. You can start by attaching A MINIMAL report design that exposes the problem. 

**A minimal report design is a report design that has very little content but shows the problem clearly**. Once we have such a design, we can advise you or identify an issue. I emphasize MINIMAL because we will not wade through a complicated report design. If you attach a large and complicated design, we will ignore that and again ask you to post a minimal design. Forgive me for repeating myself, but since it is urgent, we must make sure not to waste time getting the things we need.

**Reply by hvbtup** [reply]
As a workaround, you could create a second report (meant for PDF output) where you revert the order of the columns as intended.
This is 10 minutes work per report: Note that the editor allows you to cut and copy table columns (unless cells are merged across columns).

With several hours (or days) work you could be able to do this "manual column rearrangement" with the DE API instead, even on the fly with scripting when the report is loaded.

Another possible workaround could be to add the column header to the left *and* to the right and to use a visibility expression to decide which one should be shown.

Besides: Is there any specification saying *exactly* how RTL output should work? You only mention tables.
- How does it depend on the locale settings used in the report designer?
- What about grids?
- What about cross tabs?

Even if this could be changed in the BIRT code (and I really doubt this could be done quickly), wouldn't this be a breaking change?

**Reply by prshnt0071** [reply]
> That is fine. I thought I ask because you said it is urgent and the complete project is at risk.
> 
> For anyone to help you. We need more information. You can start by attaching A MINIMAL report design that exposes the problem.
> 
> **A minimal report design is a report design that has very little content but shows the problem clearly**. Once we have such a design, we can advise you or identify an issue. I emphasize MINIMAL because we will not wade through a complicated report design. If you attach a large and complicated design, we will ignore that and again ask you to post a minimal design. Forgive me for repeating myself, but since it is urgent, we must make sure not to waste time getting the things we need.

@wimjongman As requested, attaching the .rptdesign and the HTML content that is used as the data and displayed in the report. The report reads the same HTML content and shows the correct format with RTL when exported in docx; however, it is distorted like I mentioned when exported in PDF.
Let me know if more details are needed. I would continue to seek valuable help from everyone.

Report Design: 
[Request_Ar.zip](https://github.com/eclipse/birt/files/10057859/Request_Ar.zip)

HTML Content:
[HTML Content.txt](https://github.com/eclipse/birt/files/10057864/HTML.Content.txt)

**Reply by speckyspooky** [reply]
I took a look into the report.
But currently there are different task before we can investgate.

point 1): I figured out that you use BIRT 4.9 or BIRT 4.10.
=> So I started with BIRT 4.10. - Which version do you use?

point 2) :The next thing is that you have a data source added (your data source) which cannot be reached.
=> So I added another exampke dsBasic3 with the demo-data of "classic cars" to get data (rows).

point 3): The report have a unreachable library "omniReports_library", so we cannot say which impact it has.
=> So I ignored the library and started the report.

Based on your report with the 3 decisions I cannot rebuilt your effect. The HTML and PDF-output is fine.

Attached you will get you report with my changes and my PDF result.
Can you please check the details of your version and verify the result of my changed report.

[Request_Ar_with_classicCar.zip](https://github.com/eclipse/birt/files/10059993/Request_Ar_with_classicCar.zip)

[Request_Ar_with_classicCar.pdf](https://github.com/eclipse/birt/files/10059994/Request_Ar_with_classicCar.pdf)

**Reply by prshnt0071** [reply]
@speckyspooky Thanks for a quick help. Apologies for not reverting as was not keeping well.
Just to answer your above queries -
1. We are using BIRT 4.10
2. Can you add the HTML content attached as a demo data someway and try to generate that data in the BIRT report?
3. Yes, you can ignore "omniReports_library" as it serves the purpose and limited to only applying page numbers to our report   

I did see your report but I would request if the attached HTML content could be generated in the report. This will give us an actual effect and replicate the problem. Then maybe more feedback from you'll could be posted.  

Note: I have also realized that the text when generated in the PDF are not getting wrapped, they are displayed being cropped.

**Reply by speckyspooky** [reply]
@prshnt0071
My questions are answered and with your hint of the HTML-text I figured out your main issue and also the solution.
Or better an suggestion to solve partial your problem.

The main topic is that the HTML-emitter and PDF-emitter works different because the browser interprets your report result and your HTML which is included in an HTML-text-element. But the PDF-emitter works like desinged in the past.

The main difference is that for PDF you must set the report-orientation of "Right to Left" - see screens attached.
I tested it with your report and your HTML-exapmle. Attached you will find the new report version of RTL and the according PDF result.

The only think which is a little bit confusing is the order of the table columns.
Because you see it on my screens the columns will be inverted in the order of the result.

### **screen 01 / report setting RTL**
![2022-11-23 20_51_18-workspace - C__Users_gutma_Downloads_Request_Ar_Request_Ar_with_classicCar rptde](https://user-images.githubusercontent.com/41593722/203636154-526121d8-e231-4511-9a77-8718b80badf6.png)

### **screen 02 / report result sequence**
![2022-11-23 20_47_33-BIRT Report Viewer – Mozilla Firefox](https://user-images.githubusercontent.com/41593722/203636172-0dc1fe53-0994-4b6c-b1fb-4bddf628e3b9.png)

### **RTL report version**
[Request_Ar_with_classicCar-RTL.zip](https://github.com/eclipse/birt/files/10078825/Request_Ar_with_classicCar-RTL.zip)

### **PDF result**
[Request_Ar_with_classicCar-RTL.pdf](https://github.com/eclipse/birt/files/10078826/Request_Ar_with_classicCar-RTL.pdf)

**Reply by prshnt0071** [reply]
@speckyspooky Thank you very much for the steps. But I have a question -

The output that I see is without the row label that was present in the original .rptdesign file as seen below circled in blue.
![image](https://user-images.githubusercontent.com/118506393/203965898-29d41a49-a0eb-4f06-b32c-fb1d7e725e31.png)

However, I forgot to mention one thing that the report that gives me problem when exported to PDF, works well when exported to DOCX. So, if I change the orientation to RTL will that not create problem for report when exported to DOCX.

Does this mean that, I have to create two reports - one to work for PDF and another to work for DOCX?

**Reply by speckyspooky** [reply]
@prshnt0071 It isn't necessary to create 2 reports. You can use the same report with RTL-orientation für PDF and DOCX.

The good thing is that the DOCX-emitter gets more information from the report-layout-attributes like the PDF-emitter.
The pdf-emiter ignore your RTL-attribute on table level. These won't be copied in to to the PDF-structure.
Only the RTL-orientation on top level is deploeyed to the PDF-document.

**Reply by prshnt0071** [reply]
@speckyspooky Thanks a lot for your valuable inputs. We created a separate report for PDF with Orientation as RTL and were able to achieve what was desired.

However, we are facing below issues when the report is extracted in PDF -
1. The same HTML content are not uniform in the DOCX as well as in PDF. Please see below -
![Untitled](https://user-images.githubusercontent.com/118506393/205066060-027071a8-9772-4d0c-a66a-6d435c4803eb.png)

2. Unexpected page break happening in the PDF output which does not occur in the DOCX
Note: The Layout = 'Auto' in our case
![Untitled1](https://user-images.githubusercontent.com/118506393/205068200-1515a7b7-1e50-42ce-8227-73fe04567020.png)

@All - Thanking you all in advance!! I would wait for your expertise inputs.

**Reply by prshnt0071** [reply]
Dear All, Can anyone lend their help on the above problem? I trust that someone from our community must have faced similar problem previously.

**Reply by wimjongman** [reply]
The original discussion has been resolved. Please do not follow up with additional questions in the same discussion.

**Reply by prshnt0071** [reply]
@wimjongman Noted. But can I open another thread for getting answers to the additional questions?

**Reply by wimjongman** [reply]
Sure.

**Comment by wimjongman** [comment, marked as answer]
> @speckyspooky Thanks a lot for your valuable inputs. We created a separate report for PDF with Orientation as RTL and were able to achieve what was desired.
